### PR TITLE
Add navigatehistory controller

### PR DESCRIPTION
**Pull request by Rinder5**

### CODE REVIEW

**Comment by Rinder5** [comment]
redid up/down under a new branch since a lot of merge conflicts happened since then.

File: src/main/java/seedu/toluist/dispatcher/Dispatcher.java.rej [hunk 1]
```diff
@@ -0,0 +1,42 @@
+diff a/src/main/java/seedu/toluist/dispatcher/Dispatcher.java b/src/main/java/seedu/toluist/dispatcher/Dispatcher.java	(rejected hunks)
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/dispatcher/Dispatcher.java.rej]: delete this file lol

Comment (Melvin-Tan) [reply src/main/java/seedu/toluist/dispatcher/Dispatcher.java.rej]: or handle the merge conflict

Comment (Rinder5) [reply src/main/java/seedu/toluist/dispatcher/Dispatcher.java.rej]: =P

File: src/main/java/seedu/toluist/model/CommandHistoryList.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    private ArrayList<String> commandHistory;
+    private int historyPointer = 0;
+    private String currentCommand;
+
+    public CommandHistoryList() {
+        commandHistory = new ArrayList<String>();
+        currentCommand = "";
+    }
+
+    public ArrayList<String> getAsArrayList() {
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/model/CommandHistoryList.java]: change to getCommandHistoryList or getListOfCommandHistory or getCommandHistory?

File: src/main/java/seedu/toluist/controller/HistoryController.java [hunk 3]
```diff
@@ -18,19 +18,19 @@
...
 
     public void execute(String command) {
         logger.info(getClass().getName() + " will handle command");
 
-        String result = String.join("\n", commandHistory);
+        String result = String.join("\n", commandHistory.getAsArrayList());
 
         uiStore.setCommandResult(new CommandResult(String.format(
-                RESULT_MESSAGE, result, StringUtil.nounWithCount("command", commandHistory.size()))));
+                RESULT_MESSAGE, result, StringUtil.nounWithCount("command", commandHistory.getAsArrayList().size()))));
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/controller/HistoryController.java]: commandHistory.getAsArrayList() is called twice so can consider making it a variable first.
List<String> commandHistoryList = commandHistory.get.....

File: src/main/java/seedu/toluist/model/CommandHistoryList.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        if (historyPointer != 0) {
+            historyPointer--;
+        }
+
+        return commandHistory.get(historyPointer);
+    }
+
+    public int size() {
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/model/CommandHistoryList.java]: I think this is not needed actually....

File: src/test/java/guitests/NavigateHistoryCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    public void viewPreviousThenCurrentAfterACommandAndAHalf() {
+        String command1 = "history";
+        commandBox.runCommand(command1);
+        String command2 = "hi";
+        commandBox.enterCommand(command2);
+        mainGui.press(KeyCode.UP);
+        mainGui.press(KeyCode.DOWN);
+        assertTrue(UiStore.getInstance().getObservableCommandInput().getValue().getCommand() == command2);
+    }
+}
```
Comment (Melvin-Tan) [inline src/test/java/guitests/NavigateHistoryCommandTest.java]: try failure cases and test what happens
Lets say i have 2 commands already
then i do up up up, down down down

File: src/main/java/seedu/toluist/model/CommandHistoryList.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package seedu.toluist.model;
```
Comment (louislai) [inline src/main/java/seedu/toluist/model/CommandHistoryList.java:2]: Please add unit tests for this class

File: src/main/java/seedu/toluist/ui/MainWindow.java [hunk 7]
```diff
@@ -126,10 +127,20 @@ private void configureUndoKeyCombination() {
         FxViewUtil.setKeyCombination(getRoot(), keyCombination, handler);
     }
 
+    private void configureHistoryNavigationKeyPresses() {
+        String[] keyNames = new String[] { "up", "down"};
+        Arrays.stream(keyNames).forEach(keyName -> {
+            KeyCode keycode = getKeyCode(keyName);
+            String navigateCommand = "navigatehistory " + keyName;
+            EventHandler<ActionEvent> handler = event -> dispatcher.dispatch(navigateCommand);
+            FxViewUtil.setKeyCode(getRoot(), keycode, handler);
```
Comment (louislai) [inline src/main/java/seedu/toluist/ui/MainWindow.java]: shouldnt this only be added to the command box, not the whole app?

File: src/main/java/seedu/toluist/ui/MainWindow.java [hunk 8]
```diff
@@ -126,10 +127,20 @@ private void configureUndoKeyCombination() {
         FxViewUtil.setKeyCombination(getRoot(), keyCombination, handler);
     }
 
+    private void configureHistoryNavigationKeyPresses() {
+        String[] keyNames = new String[] { "up", "down"};
+        Arrays.stream(keyNames).forEach(keyName -> {
+            KeyCode keycode = getKeyCode(keyName);
+            String navigateCommand = "navigatehistory " + keyName;
```
Comment (Melvin-Tan) [inline src/main/java/seedu/toluist/ui/MainWindow.java]: magic strings, "up", "down", "navigatehistory" should all go to private static final SOMETHING_SOMETHING respectively..

File: src/main/java/seedu/toluist/ui/MainWindow.java (hunk 7 above)
Comment (Rinder5) [reply src/main/java/seedu/toluist/ui/MainWindow.java]: i think its better this way. we don't have any clickable thing, so the only thing the user should be able to interact with anyway is command box

Comment (Rinder5) [reply src/main/java/seedu/toluist/ui/MainWindow.java]: imo, even without focus on the command box, even if they type it should appear there

**Comment by louislai** [comment]
open the app. press up immediately. Observe the crash

File: src/main/java/seedu/toluist/ui/MainWindow.java (hunk 7 above)
Comment (Rinder5) [reply src/main/java/seedu/toluist/ui/MainWindow.java]: but i can change if you want

Comment (louislai) [reply src/main/java/seedu/toluist/ui/MainWindow.java]: thats provided you dont want to provide user with the ability to use up/down to navigate the task list later on

**Comment by louislai** [comment]
Handle your corner cases properly
- Start the app. Type a command. What should happen when i press up.
- Start the app. Type a command. What should happen when i press down.
- Type a command. Press up, then down. What happens when i press down again
- Type a command and enter it. Type another command, Press up. edit the command. What should happen when i press down then up
and any other corner cases.
new corner cases that get handled should always be supported with tests to prevent regression

File: src/test/java/guitests/NavigateHistoryCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        String command1 = "history";
+        commandBox.runCommand(command1);
+        String command2 = "hi";
+        commandBox.enterCommand(command2);
+        mainGui.press(KeyCode.UP);
+        mainGui.press(KeyCode.UP);
+        mainGui.press(KeyCode.UP);
+        mainGui.press(KeyCode.UP);
+        mainGui.press(KeyCode.UP);
+        assertTrue(uiStore.getObservableCommandInput().getValue().getCommand() == command1);
```
Comment (louislai) [inline src/test/java/guitests/NavigateHistoryCommandTest.java]: pretty sure the preferred to compare String is using .equals

Comment (louislai) [reply src/test/java/guitests/NavigateHistoryCommandTest.java]: and there is `assertEquals`

File: src/test/java/guitests/NavigateHistoryCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        assertTrue(uiStore.getObservableCommandInput().getValue().getCommand() == command2);
+    }
+
+    @Test
+    public void viewEverythingTooManyTimes() {
+        String command1 = "history";
+        commandBox.runCommand(command1);
+        String command2 = "hi";
+        commandBox.enterCommand(command2);
+        mainGui.press(KeyCode.UP);
```
Comment (louislai) [inline src/test/java/guitests/NavigateHistoryCommandTest.java]: multiple lines of code with the same structure. Refactor time

File: src/main/java/seedu/toluist/ui/MainWindow.java (hunk 8 above)
Comment (Rinder5) [reply src/main/java/seedu/toluist/ui/MainWindow.java]: @louislai this one's in the mainwindow. want me to do it for your undo and switch hotkeys too?

Comment (louislai) [reply src/main/java/seedu/toluist/ui/MainWindow.java]: can

**Comment by Rinder5** [comment]
k, addressed everything i think

File: src/test/java/seedu/toluist/model/CommandHistoryListTest.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
+//@@author A0162011A
+package seedu.toluist.model;
+
+import static org.junit.Assert.assertEquals;
+import java.util.ArrayList;
+
+import org.junit.Test;
+
+/**
+ * Tests for Tag model
```
Comment (louislai) [inline src/test/java/seedu/toluist/model/CommandHistoryListTest.java]: tag model?

**Comment by louislai** [comment]
merge master and check that nothing breaks

File: src/test/java/seedu/toluist/model/CommandHistoryListTest.java (hunk 11 above)
Comment (Rinder5) [reply src/test/java/seedu/toluist/model/CommandHistoryListTest.java]: *facepalm*

**Comment by Rinder5** [comment]
yup, looks fine. unless i'm doing it wrongly

**Comment by louislai** [comment]
i mean merge then push

**Comment by Rinder5** [comment]
looks fine...maybe?

**Comment by louislai** [comment]
add ur author claim as well

**Comment by Rinder5** [comment]
i thought i did. what part did i miss?

File: src/main/java/seedu/toluist/commons/util/FxViewUtil.java [hunk 12]
```diff
@@ -72,4 +73,19 @@ public static void setKeyCombination(Node node, KeyCombination keyCombination,
         });
     }
 
+    /**
```
Comment (louislai) [inline src/main/java/seedu/toluist/commons/util/FxViewUtil.java]: author for this

Comment (Rinder5) [reply src/main/java/seedu/toluist/commons/util/FxViewUtil.java]: do i add for that? its basically your code with different variables

File: src/main/java/seedu/toluist/ui/MainWindow.java [hunk 13]
```diff
@@ -121,15 +127,24 @@ private void configureSwitchTabKeyCombinations() {
      */
     private void configureUndoKeyCombination() {
         KeyCombination keyCombination = new KeyCodeCombination(KeyCode.Z, KeyCombination.CONTROL_DOWN);
-        String undoCommand = "undo";
+        String undoCommand = COMMAND_UNDO;
         EventHandler<ActionEvent> handler = event -> dispatcher.dispatch(undoCommand);
         FxViewUtil.setKeyCombination(getRoot(), keyCombination, handler);
     }
```
Comment (louislai) [inline src/main/java/seedu/toluist/ui/MainWindow.java:134]: author

File: src/main/java/seedu/toluist/commons/util/FxViewUtil.java (hunk 12 above)
Comment (louislai) [reply src/main/java/seedu/toluist/commons/util/FxViewUtil.java]: its another method

**Comment by louislai** [comment]
build seems to be failing

**Comment by Rinder5** [comment]
it seems to be working on my end

**Comment by louislai** [comment]
need to work on both ur end and travis https://travis-ci.org/CS2103JAN2017-W13-B3/main/builds

**Comment by Rinder5** [comment]
so we need to wait until it's done to see?

**Comment by louislai** [comment]
u can see that the tests in the previous build are alr failing https://travis-ci.org/CS2103JAN2017-W13-B3/main/builds/215960216. As for waiting to see it, you'll get used to it lol

**Comment by Rinder5** [comment]
the previous ones were from linting issues though. but yea found out where i messed up =P

**Review by louislai — APPROVED** [review]
